VSF1 instance guide (#119)
moovweb-docs/xdn-docs: guides/vsf1_instance.md

# Adding XDN to Vue Storefront Instance guide

## 1. Install Vue Storefront

> Install VSF application using this guide: https://docs.vuestorefront.io/guide/installation/linux-mac.html

Reviewer: ```suggestion
Install the Vue Storefront application using this guide: https://docs.vuestorefront.io/guide/installation/linux-mac.html
```

<b>Note</b>: XDN requires Node version >= 12, so before the installation it's recommended to run:
```bash
nvm use 12
```

## 2. Prepare VSF files for XDN

- In the new VSF project, go to `src/themes/default` (or any theme you're using) and remove `.git` folder from it to save that in Git VCS.
- Go to `.gitignore` file and remove `config/local.json` line to keep it tracked.

## 3. Install XDN

> - Install XDN packages: `yarn add -D @xdn/cli && yarn add @xdn/core @xdn/prefetch @xdn/devtools`

Reviewer 2: Shouldn't we suggest using `xdn init` instead of installing the individual packages/files manually?

Author: I think that's a kind of custom case where we need to configure that manually

Reviewer 2: Perhaps there is a bug in the CLI since ideally this should work. Trying it myself I get the following error:
```
vue-storefront git:(master) ✗ xdn init
ℹ installing @xdn/core, @xdn/cli, @xdn/prefetch, @xdn/devtools...
yarn add v1.19.1
error Running this command will add the dependency to the workspace root rather than the workspace itself, which might not be what you want - if you really meant it, make it explicit by running this command again with the -W flag (or --ignore-workspace-root-check).
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
✖ installing @xdn/core, @xdn/cli, @xdn/prefetch, @xdn/devtools... failed.
Error: Process exited with code 1
    at ChildProcess.<anonymous> (/Users/tristanlee/.nvm/versions/node/v12.18.4/lib/node_modules/@xdn/cli/utils/run.js:30:16)
    at ChildProcess.emit (events.js:315:20)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
Error: Process exited with code 1
    at ChildProcess.<anonymous> (/Users/tristanlee/.nvm/versions/node/v12.18.4/lib/node_modules/@xdn/cli/utils/run.js:30:16)
    at ChildProcess.emit (events.js:315:20)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
```

Reviewer 2: Actually, even running the package install as you have documented I still get the error:
```
vue-storefront git:(master) ✗ yarn add -D @xdn/cli && yarn add @xdn/core @xdn/prefetch @xdn/devtools
yarn add v1.19.1
error Running this command will add the dependency to the workspace root rather than the workspace itself, which might not be what you want - if you really meant it, make it explicit by running this command again with the -W flag (or --ignore-workspace-root-check).
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

If this is the correct place to install, I think we need to document the `-W` flag as well.

Author: Oh, yea, good catch! It's a `-W` flag should be added

Author: just added

> - Next steps will cover the XDN key files you need to add. For more information about them see https://developer.moovweb.com/guides/starter

Reviewer: ```suggestion
```

I don't like sending SPA developers to the starter guide. It's a rabbit hole and I don't want to confuse them.  I think you can just continue on with the steps here.

> - Add `xdn.config.js` file with your origin and images hosts. For example:

Reviewer: ```suggestion
- Create a file called `xdn.config.js` in the root directory of your project and configure your origin and images hosts as backends. For example:
```

```js
module.exports = {
  routes: './xdn/routes.js',
  backends: {
    origin: {
      domainOrIp: 'demo.vuestorefront.io',
      hostHeader: 'demo.vuestorefront.io',
    },
  },
  server: {
    path: './dist-xdn-server/server.js',
  },
  includeNodeModules: true,
};
```

> - Add `xdn/service-worker.js` file:

Reviewer: ```suggestion
- Create `xdn/service-worker.js`, which will contain the source code for your service worker:
```

```js
import { Prefetcher } from "@xdn/prefetch/sw";
import { clientsClaim, skipWaiting } from "workbox-core";
import { precacheAndRoute } from 'workbox-precaching';
import DeepFetchPlugin, { DeepFetchCallbackParam } from '@xdn/prefetch/sw/DeepFetchPlugin'
import { prefetch } from "@xdn/prefetch/window/prefetch";
skipWaiting();
clientsClaim();
precacheAndRoute(self.__WB_MANIFEST || []);
new Prefetcher().route()
```

> - Add `xdn/browser.js` file. For example:

Reviewer: ```suggestion
- Create `xdn/browser.js` with the following content:
```

```js
import installDevtools from "@xdn/devtools/install";
import install from "@xdn/prefetch/window/install";

> document.addEventListener("DOMContentLoaded", () => {

Reviewer: ```suggestion

document.addEventListener("DOMContentLoaded", () => {
```

  console.info("[XDN browser] DOMContentLoaded -> running install()");
  install({
    forcePrefetchRatio: 0.5, // forcely prefetch 50% of non-cached content for higher hit rate
  });
  console.info("[XDN browser] DOMContentLoaded -> running installDevtools()");
  installDevtools();
});
```

> - Add `xdn/routes.js` file. For example:

Reviewer: ```suggestion
- Create `xdn/routes.js`. Here you will configure caching for each route in your application. Here is an example:
```

  proxy('origin');
});
router.get(`/api/stock/${SPLAT}`, ({ proxy, cache }) => {
  cache(CACHE_PAGES);
  proxy('origin');
});
// pages
pages.forEach(page => {
  router.get(page, ({ cache, proxy }) => {
    cache(CACHE_PAGES);
    proxy(BACKENDS.js);
  });
});
// fallback
router.fallback(({ proxy }) => {
  proxy(BACKENDS.js);
});
export default router;
```
- Add `xdn/cache.js` file. For example:
```js
const TIME_1H = 60 * 60;
const TIME_4H = TIME_1H * 4;
const TIME_1D = TIME_1H * 24;
/**
 * The default cache setting for pages in the shopping flow
 */
export const CACHE_PAGES = {
  edge: {
    maxAgeSeconds: TIME_4H,
    forcePrivateCaching: true,
    staleWhileRevalidateSeconds: TIME_1H, // this way stale items can still be prefetched
  },
  browser: {
    maxAgeSeconds: TIME_4H,
    serviceWorkerSeconds: TIME_4H,
    spa: true,
  },
};
/**
 * The default cache setting for static assets like JS, CSS, and images.
 */
export const CACHE_ASSETS = {
  edge: {
    maxAgeSeconds: TIME_1D,
    forcePrivateCaching: true,
    staleWhileRevalidateSeconds: TIME_1H, // this way stale items can still be prefetched
  },
  browser: {
    maxAgeSeconds: TIME_1D,
    serviceWorkerSeconds: TIME_1D,

>     spa: true,

Reviewer: Are you sure you need this? If so, why?

Author: @markbrocato I'm using that since it's an SPA app

  },
};
```

> - Add `xdn/webpack.xdn.client.config.js` file:

Reviewer: ```suggestion
- Create `xdn/webpack.xdn.client.config.js`.  This webpack configuration will bundle the service worker and the code to install it when the app loads in the browser.
```

```js
const path = require('path')
const webpack = require('webpack')
module.exports = {
  entry: {
    browser: './xdn/browser.js',
    'service-worker': './xdn/service-worker.js',
  },
  mode: 'production',
  resolve: {
    extensions: ['.js'],
  },
  output: {
    filename: '[name].js',
    path: path.resolve(__dirname, '../dist-xdn'),
  },
  plugins: [
    new webpack.DefinePlugin({
      'process.env.DEBUG_SW': JSON.stringify(process.env.DEBUG_SW || false),
    }),
  ],
}
```
- Add `core/build/webpack.xdn.config.ts` file:
```ts
import { buildLocaleIgnorePattern } from '@vue-storefront/i18n/helpers';
import path from 'path';
import fs from 'fs';
import CaseSensitivePathsPlugin from 'case-sensitive-paths-webpack-plugin';
import VueLoaderPlugin from 'vue-loader/lib/plugin';
import autoprefixer from 'autoprefixer';
import webpack from 'webpack';
import dayjs from 'dayjs';
import config from 'config';
// const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
// eslint-disable-next-line import/first

> import themeRoot from './theme-path';

Reviewer: ```suggestion
import themeRoot from './theme-path';

```
